Source/Core/VideoCommon/TextureConverterShaderGen.cpp

```diff
@@ -0,0 +1,295 @@
+
+ShaderCode GenerateTextureConverterShaderCode(APIType api_type,
+                                              const convertion_shader_uid_data* uid_data)
+{
+  ShaderCode out;
+
+  std::array<float, 28> colmat = {};
+  float* const const_add = &colmat[16];
+  float* const color_mask = &colmat[20];
+  color_mask[0] = color_mask[1] = color_mask[2] = color_mask[3] = 255.0f;
+  color_mask[4] = color_mask[5] = color_mask[6] = color_mask[7] = 1.0f / 255.0f;
+
+  if (api_type == APIType::OpenGL)
+    out.Write("SAMPLER_BINDING(9) uniform sampler2DArray samp9;\n"
+              "#define samp0 samp9\n"
+              "#define uv0 f_uv0\n"
+              "in vec3 uv0;\n"
+              "out vec4 ocol0;\n");
+
+  else if (api_type == APIType::Vulkan)
+    out.Write("SAMPLER_BINDING(0) uniform sampler2DArray samp0;\n"
+              "layout(location = 0) in vec3 uv0;\n"
+              "layout(location = 1) in vec4 col0;\n"
+              "layout(location = 0) out vec4 ocol0;");
+
+  bool mono_depth = uid_data->is_depth_copy && g_ActiveConfig.bStereoEFBMonoDepth;
+  out.Write("void main(){\n"
+            "  vec4 texcol = texture(samp0, %s);\n",
+            mono_depth ? "vec3(uv0.xy, 0.0)" : "uv0");
+
+  if (uid_data->is_depth_copy)
+  {
+    switch (uid_data->dst_format)
+    {
+    case EFBCopyFormat::R4:  // Z4
+      colmat[3] = colmat[7] = colmat[11] = colmat[15] = 1.0f;
+      break;
+    case EFBCopyFormat::R8_0x1:  // Z8
+    case EFBCopyFormat::R8:      // Z8H
+      colmat[0] = colmat[4] = colmat[8] = colmat[12] = 1.0f;
+      break;
+
+    case EFBCopyFormat::RA8:  // Z16
+      colmat[1] = colmat[5] = colmat[9] = colmat[12] = 1.0f;
+      break;
+
+    case EFBCopyFormat::RG8:  // Z16 (reverse order)
+      colmat[0] = colmat[4] = colmat[8] = colmat[13] = 1.0f;
+      break;
+
+    case EFBCopyFormat::RGBA8:  // Z24X8
+      colmat[0] = colmat[5] = colmat[10] = 1.0f;
+      break;
+
+    case EFBCopyFormat::G8:  // Z8M
+      colmat[1] = colmat[5] = colmat[9] = colmat[13] = 1.0f;
+      break;
+
+    case EFBCopyFormat::B8:  // Z8L
+      colmat[2] = colmat[6] = colmat[10] = colmat[14] = 1.0f;
+      break;
+
+    case EFBCopyFormat::GB8:  // Z16L - copy lower 16 depth bits
+      // expected to be used as an IA8 texture (upper 8 bits stored as intensity, lower 8 bits
+      // stored as alpha)
+      // Used e.g. in Zelda: Skyward Sword
+      colmat[1] = colmat[5] = colmat[9] = colmat[14] = 1.0f;
+      break;
+
+    default:
+      ERROR_LOG(VIDEO, "Unknown copy zbuf format: 0x%X", static_cast<int>(uid_data->dst_format));
+      colmat[2] = colmat[5] = colmat[8] = 1.0f;
+      break;
+    }
+  }
+  else if (uid_data->is_intensity)
+  {
+    const_add[0] = const_add[1] = const_add[2] = 16.0f / 255.0f;
+    switch (uid_data->dst_format)
+    {
+    case EFBCopyFormat::R4:      // I4
+    case EFBCopyFormat::R8_0x1:  // I8
+    case EFBCopyFormat::R8:      // IA4
+    case EFBCopyFormat::RA4:     // IA8
+    case EFBCopyFormat::RA8:     // I8
```

Comment:
Comments here are wrong. R8 should be I8, RA4 should be IA4, and RA8 should be IA8.

Reply:
Please address this comment.

Reply:
Please note that since writing this review, I have written a patch which corrected these comments and had it merged to master.